### PR TITLE
[ASTS] Implementation of SweepBucketsTable

### PR DESCRIPTION
**Pull request by mdaudali**
## General
**Before this PR**:
Breaking up https://github.com/palantir/atlasdb/pull/7313 into individual PRs representing each table. They will all be in the same class.

We have various interfaces representing logical tables necessary for tracking bucket state, but no concrete implementation.
**After this PR**:
SweepBucketPointerTable

==COMMIT_MSG==
==COMMIT_MSG==

**Priority**: P2

**Concerns / possible downsides (what feedback would you like?)**:
Not for this one.
## Testing and Correctness
**What, if any, assumptions are made about the current state of the world? If they change over time, how will we find out?**:
N/A
**What was existing testing like? What have you done to improve it?**:
Added tests
**If this PR contains complex concurrent or asynchronous code, is it correct? The onus is on the PR writer to demonstrate this.**:
N/A

## Execution
Not wired up.

## Scale
**Would this PR be expected to pose a risk at scale? Think of the shopping product at our largest stack.**:
No
**Would this PR be expected to perform a large number of database calls, and/or expensive database calls (e.g., row range scans, concurrent CAS)?**:
No
**Would this PR ever, with time and scale, become the wrong thing to do - and if so, how would we know that we need to do something differently?**:
Not this one
## Development Process
**Where should we start reviewing?**:
DBA
**If this PR is in excess of 500 lines excluding versions lock-files, why does it not make sense to split it?**:
N/A
**Please tag any other people who should be aware of this PR**:
@jeremyk-91
@sverma30
@raiju

### CODE REVIEW

File: atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/sweep/asts/bucketingthings/DefaultSweepAssignedBucketStore.java [hunk 1]
```diff
@@ -174,6 +185,32 @@ public void updateStartingBucketForShardAndStrategy(Bucket newStartingBucket) {
...
+        Cell cell = SweepAssignedBucketStoreKeyPersister.INSTANCE.sweepBucketsCell(bucket);
+        casCell(
+                cell,
+                oldTimestampRange.map(timestampRangePersister::trySerialize),
+                timestampRangePersister.trySerialize(newTimestampRange));
+    }
+
+    @Override
+    public void deleteBucketEntry(Bucket bucket) {
+        throw new UnsupportedOperationException("deleteBucketEntry is not implemented yet.");
```
Comment (jeremyk-91) [inline atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/sweep/asts/bucketingthings/DefaultSweepAssignedBucketStore.java:222]: question: we need #7197 or similar for this right?

File: atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/sweep/asts/bucketingthings/DefaultSweepAssignedBucketStore.java [hunk 2]
```diff
@@ -174,6 +185,32 @@ public void updateStartingBucketForShardAndStrategy(Bucket newStartingBucket) {
         }
     }
 
+    @Override
+    public Set<SweepableBucket> getSweepableBuckets(Set<Bucket> startBuckets) {
+        List<SweepAssignedBucketsRow> rows = startBuckets.stream()
+                .flatMap(bucket -> Stream.of(
+                        SweepAssignedBucketStoreKeyPersister.INSTANCE.sweepBucketsRow(bucket),
+                        SweepAssignedBucketStoreKeyPersister.INSTANCE.nextSweepBucketsRow(bucket)))
```
Comment (jeremyk-91) [inline atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/sweep/asts/bucketingthings/DefaultSweepAssignedBucketStore.java:204]: Might be worth a comment here explaining the logic of choosing two rows.

File: atlasdb-impl-shared/src/test/java/com/palantir/atlasdb/sweep/asts/bucketingthings/DefaultSweepAssignedBucketStoreTest.java [hunk 3]
```diff
@@ -162,4 +170,106 @@ public void updateStartingBucketForShardFailsAfterTooManyAttempts() {
...
+                bucket -> store.putTimestampRangeForBucket(bucket.bucket(), Optional.empty(), bucket.timestampRange()));
+
+        assertThat(store.getSweepableBuckets(Set.of(Bucket.of(shardAndStrategy, 100))))
+                .containsExactlyInAnyOrderElementsOf(buckets.subList(100, 200));
+    }
+
+    @Test
+    public void getSweepableBucketsReturnsLessThanMaxBucketLimitIfLessThanTwoHundredBucketsPresent() {
+        ShardAndStrategy shardAndStrategy = ShardAndStrategy.of(12, SweeperStrategy.THOROUGH);
+        List<SweepableBucket> buckets = IntStream.range(0, 10)
```
Comment (jeremyk-91) [inline atlasdb-impl-shared/src/test/java/com/palantir/atlasdb/sweep/asts/bucketingthings/DefaultSweepAssignedBucketStoreTest.java]: optional: should we parameterize on the number and get `0`, `1`, `12` and `199` buckets? Might be overkill

File: atlasdb-impl-shared/src/test/java/com/palantir/atlasdb/sweep/asts/bucketingthings/DefaultSweepAssignedBucketStoreTest.java [hunk 4]
```diff
@@ -162,4 +170,106 @@ public void updateStartingBucketForShardFailsAfterTooManyAttempts() {
...
+        Bucket bucket = Bucket.of(ShardAndStrategy.of(12, SweeperStrategy.THOROUGH), 512);
+        TimestampRange newTimestampRange = TimestampRange.of(1, 2);
+
+        store.putTimestampRangeForBucket(bucket, Optional.empty(), newTimestampRange);
+        Set<SweepableBucket> sweepableBuckets = store.getSweepableBuckets(Set.of(bucket));
+        assertThat(sweepableBuckets).containsExactly(SweepableBucket.of(bucket, newTimestampRange));
+    }
+
+    @Test
+    @Disabled // TODO(mdaudali): Deletion is not implemented yet
```
Comment (jeremyk-91) [inline atlasdb-impl-shared/src/test/java/com/palantir/atlasdb/sweep/asts/bucketingthings/DefaultSweepAssignedBucketStoreTest.java:264]: 👍 nice!

**Review by jeremyk-91 — APPROVED** [review]
👍 Looks good, thanks!

File: atlasdb-impl-shared/src/test/java/com/palantir/atlasdb/sweep/asts/bucketingthings/DefaultSweepAssignedBucketStoreTest.java (hunk 3 above)
Comment (mdaudali) [reply atlasdb-impl-shared/src/test/java/com/palantir/atlasdb/sweep/asts/bucketingthings/DefaultSweepAssignedBucketStoreTest.java]: Ha, sure, will do!

File: atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/sweep/asts/bucketingthings/DefaultSweepAssignedBucketStore.java (hunk 2 above)
Comment (mdaudali) [reply atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/sweep/asts/bucketingthings/DefaultSweepAssignedBucketStore.java:204]: Yeah I have a long comment I'll push up later, it got lost in the exercise of splitting all these PRs up at speed.

File: atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/sweep/asts/bucketingthings/DefaultSweepAssignedBucketStore.java (hunk 1 above)
Comment (mdaudali) [reply atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/sweep/asts/bucketingthings/DefaultSweepAssignedBucketStore.java:222]: Yes

**Comment by mdaudali** [comment]
👍

**Pull request merged**